### DISCUSSION TITLE
model prediction returns "None" for 'Keypoints', 'probs' and 'masks' attributes in YOLOV8

### DISCUSSION THREAD

**Discussion by anemicwolf** [ultralytics/ultralytics#5970] [Q&A]
Hello, I can't seem to figure out why after running the code, results would always generate 'None' type for the 'Keypoints', 'probs' and 'masks' attributes. I'm assuming that it means there is no pointers to the respective objects or the objects were never generated to begin with. Any help would be much appreciated.

# Code
import cv2
from ultralytics import YOLO

model = YOLO('yolov8x.pt')  # load an official detection model

cap = cv2.VideoCapture(0)  # 0 represents the default webcam (you can change it if you have multiple cameras)

if not cap.isOpened():
    raise Exception("Could not open video device")

ret, frame = cap.read()

cap.release()

cv2.imwrite('captured_image.jpg', frame)

results = model('captured_image.jpg', save=True, conf=0.5, classes=0)
print(results)

# Console Output:
image 1/1 /home/anemicwolf/PycharmProjects/ImageCaptureWebcam/captured_image.jpg: 480x640 1 person, 1473.4ms
Speed: 1.3ms preprocess, 1473.4ms inference, 1.3ms postprocess per image at shape (1, 3, 480, 640)
Results saved to runs/detect/predict28
[ultralytics.engine.results.Results object with attributes:

boxes: ultralytics.engine.results.Boxes object
keypoints: None
masks: None
names: {0: 'person', 1: 'bicycle', 2: 'car', 3: 'motorcycle', 4: 'airplane', 5: 'bus', 6: 'train', 7: 'truck', 8: 'boat', 9: 'traffic light', 10: 'fire hydrant', 11: 'stop sign', 12: 'parking meter', 13: 'bench', 14: 'bird', 15: 'cat', 16: 'dog', 17: 'horse', 18: 'sheep', 19: 'cow', 20: 'elephant', 21: 'bear', 22: 'zebra', 23: 'giraffe', 24: 'backpack', 25: 'umbrella', 26: 'handbag', 27: 'tie', 28: 'suitcase', 29: 'frisbee', 30: 'skis', 31: 'snowboard', 32: 'sports ball', 33: 'kite', 34: 'baseball bat', 35: 'baseball glove', 36: 'skateboard', 37: 'surfboard', 38: 'tennis racket', 39: 'bottle', 40: 'wine glass', 41: 'cup', 42: 'fork', 43: 'knife', 44: 'spoon', 45: 'bowl', 46: 'banana', 47: 'apple', 48: 'sandwich', 49: 'orange', 50: 'broccoli', 51: 'carrot', 52: 'hot dog', 53: 'pizza', 54: 'donut', 55: 'cake', 56: 'chair', 57: 'couch', 58: 'potted plant', 59: 'bed', 60: 'dining table', 61: 'toilet', 62: 'tv', 63: 'laptop', 64: 'mouse', 65: 'remote', 66: 'keyboard', 67: 'cell phone', 68: 'microwave', 69: 'oven', 70: 'toaster', 71: 'sink', 72: 'refrigerator', 73: 'book', 74: 'clock', 75: 'vase', 76: 'scissors', 77: 'teddy bear', 78: 'hair drier', 79: 'toothbrush'}
orig_img: array([[[167, 140, 103],
        [167, 141, 101],
        [168, 139, 100],
        ...,
        [153, 148, 147],
        [154, 149, 146],
        [156, 151, 148]],

        [ 71,  55, 119],
        [ 72,  57, 119],
        [ 73,  58, 120]]], dtype=uint8)
orig_shape: (480, 640)
path: '/home/anemicwolf/PycharmProjects/ImageCaptureWebcam/captured_image.jpg'
probs: None
save_dir: 'runs/detect/predict28'
speed: {'preprocess': 1.262664794921875, 'inference': 1473.3686447143555, 'postprocess': 1.2753009796142578}]

Process finished with exit code 0

**Comment by pderrenger** [comment]
@anemicwolf hello! The `keypoints`, `probs`, and `masks` attributes returning `None` typically indicates that the model you are using is not configured to predict these specific outputs. In your case, you're using a YOLOv8 model that is likely trained for object detection, which by default provides bounding box outputs (`boxes`) but not keypoints, probabilities for classification tasks, or segmentation masks.

To obtain keypoints, you would need a model that is specifically trained for pose estimation or keypoint detection. Similarly, for `probs`, you would need a classification model, and for `masks`, a segmentation model.

If you want to perform these tasks, make sure to use the appropriate model file that supports the desired output. For example, for segmentation, you would use a model file that ends with `-seg.pt`, and for pose estimation, you would use a model trained for that purpose.

Please ensure that you are using the correct model for your task. If you are looking to perform detection, segmentation, and pose estimation simultaneously, you would need a model that is trained to handle all these tasks or use separate models for each task and combine the results accordingly.

For more information on how to use different models for various tasks such as detection, segmentation, classification, and pose estimation, you can refer to the Ultralytics documentation. 📚😊

**Reply by DuyDev1205** [reply]
thanks for support , but can you give me the example for this @@

**Reply by pderrenger** [reply]
Hello @DuyDev1205,

Thank you for your patience! To address your issue with the `keypoints`, `probs`, and `masks` attributes returning `None`, here's a more detailed example tailored to your needs.

If you want to perform segmentation or pose estimation, you need to use the appropriate model. Below are examples for both tasks:

### Segmentation Example
To use a segmentation model, you can load a model like `yolov8n-seg.pt`:

```python
import cv2
from ultralytics import YOLO

# Load a segmentation model
model = YOLO('yolov8n-seg.pt')

# Capture an image from the webcam
cap = cv2.VideoCapture(0)
if not cap.isOpened():
    raise Exception("Could not open video device")

ret, frame = cap.read()
cap.release()

cv2.imwrite('captured_image.jpg', frame)

# Perform segmentation
results = model('captured_image.jpg', save=True, conf=0.5)
print(results)

# Access masks
for result in results:
    print(result.masks)  # This should now contain segmentation masks
```

### Pose Estimation Example
For pose estimation, use a model like `yolov8n-pose.pt`:

```python
import cv2
from ultralytics import YOLO

# Load a pose estimation model
model = YOLO('yolov8n-pose.pt')

# Capture an image from the webcam
cap = cv2.VideoCapture(0)
if not cap.isOpened():
    raise Exception("Could not open video device")

ret, frame = cap.read()
cap.release()

cv2.imwrite('captured_image.jpg', frame)

# Perform pose estimation
results = model('captured_image.jpg', save=True, conf=0.5)
print(results)

# Access keypoints
for result in results:
    print(result.keypoints)  # This should now contain keypoints
```

### Classification Example
For classification tasks, ensure you are using a classification model:

```python
import cv2
from ultralytics import YOLO

# Load a classification model
model = YOLO('yolov8n-cls.pt')

# Capture an image from the webcam
cap = cv2.VideoCapture(0)
if not cap.isOpened():
    raise Exception("Could not open video device")

ret, frame = cap.read()
cap.release()

cv2.imwrite('captured_image.jpg', frame)

# Perform classification
results = model('captured_image.jpg', save=True, conf=0.5)
print(results)

# Access probabilities
for result in results:
    print(result.probs)  # This should now contain classification probabilities
```

Please ensure you are using the latest version of the Ultralytics package to avoid any compatibility issues. If the problem persists, providing a minimum reproducible example would be very helpful. You can refer to our guide here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any further questions! 😊

**Comment by DuyDev1205** [comment]
Tks for help me :3

Vào Th 2, 1 thg 7, 2024 lúc 23:10 Paula Derrenger ***@***.***>
đã viết:
